### PR TITLE
Windowed SDPA

### PR DESCRIPTION
**Pull request by gwangTT**
### Ticket
Link to Github Issue

### Problem description
Qwen2.5-VL vision model has many vision transformer layers that uses windowed attention. This means the model computes attention locally within small image patches (“windows”), rather than globally across the entire image. Some of the model variants (3B and 72B) contains a few full-attention blocks (e.g., 4 out of 32), but most attention layers operate in this windowed mode.

Using our existing SDPA for this windowed attention requires construction and transfer to device of a `attention_mask` matrix that is `seq_len` by `seq_len` in size. `seq_len` is a count of input image tokens which can be quite large -- e.g., 300 DPI scanned document in US letter size would be turned into ~43000 image tokens by Qwen2.5-VL's image processor. 

Transfer of up to `43000x43000` attention_mask matrix to device can take dozens of seconds. This is a major bottleneck of the Qwen2.5-VL model execution on TT hardware.

### What's changed
This PR introduces a kernel that is aimed to remove the need to construct (hence the need to transfer) the `attention_mask` on the host. The kernel (named `windowed_SDPA`) constructs the `attention_mask` on the fly based on a much smaller input -- a sequence of integers that defines the bounds of the windows. In Qwen2.5-VL's default settings, for a 300 DPI scanned document, the length of this sequence is ~700. 

This PR contains the kernel implementation and unit tests of the `windowed_SDPA` kernel. The compute kernel code and writer kernel code of `windowed_SDPA` are simplified copies from the SDPA kernel; the reader kernel code of `windowed_SDPA` shows significant difference than that of the SDPA kernel because this is where the `attention_mask` is constructed on the fly. 

The integration of `windowed_SDPA` into Qwen2.5-VL model code will come soon. This is one of several milestones in our planned roadmap towards optimized Qwen2.5-VL-72B with 300DPI scanned document inputs.

Performance measurements of kernel runs using `test_windowed_sdpa.py` show for 300 DPI (i.e., `seqlen = 43008`):
- ****~30x** speedup** in windowed attention, compared with previous use of SDPA
- ****~10x** speedup** in full attention, compared with previous use of SDPA
image

#### Needs minor fixing
- [x] ["Please use TensorAccessor instead."](https://github.com/tenstorrent/tt-metal/actions/runs/17002513368/job/48206769011?pr=26960#step:3:56) 
  - addressed through [Use TensorAccessor; Thanks Claude Opus 4.1](https://github.com/tenstorrent/tt-metal/pull/26960/commits/95518c898554b8ca3820c4522c4954027700817e) 
- [x] revert [[to be reverted] turn on just qwenvl unit tests](https://github.com/tenstorrent/tt-metal/pull/26960/commits/f935bc8b744e91142670e03d4b5d3acb4ddb6b13)
  - reverted through [458c4e3](https://github.com/tenstorrent/tt-metal/pull/26960/commits/458c4e37d371737fdb6ec6779ffc1fd7c4548b71) 

### Checklist
- [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
  - [x] https://github.com/tenstorrent/tt-metal/actions/runs/16999545952 
- (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
  - [x] https://github.com/tenstorrent/tt-metal/actions/runs/17010225230
  - [x] https://github.com/tenstorrent/tt-metal/actions/runs/17413968125 to check reduced timeout value
- [x] revert ec2e55864da0d437e8114ca141b7e56a5782f336 after check of reduced timeout value
  - reverted through [dfdbcd3](https://github.com/tenstorrent/tt-metal/pull/26960/commits/dfdbcd371115d7ac2a1b46533c041274a5d8184a)

### CODE REVIEW

**Comment by gwangTT** [comment]
FYI, @ppetrovicTT @mtairum @handrewsTT, in case this could be helpful for you.

**Comment by davorchap** [comment]
Can this scheme be used in GPT OSS?

**Comment by gwangTT** [comment]
> Can this scheme be used in GPT OSS?

`windowed_SDPA` is made to work with image tokens and thus causality was unnecessary (thus not implemented). The code could be enhanced to support causality, which GPT OSS requires for causal sliding window attention.

@kpaigwar @avoraTT, what do you think?

**Review by yieldthought — APPROVED** [review]

File: .github/workflows/t3000-unit-tests-impl.yaml [hunk 1]
```diff
@@ -40,7 +40,7 @@ jobs:
           { name: "t3k grok tests", arch: wormhole_b0, cmd: run_t3000_grok_tests, timeout: 30, owner_id: U03HY7MK4BT}, #Mark O'Connor
           { name: "t3k unet shallow tests", arch: wormhole_b0, cmd: run_t3000_unet_shallow_tests, timeout: 30, owner_id: U06ECNVR0EN}, #Evan Smal
           { name: "t3k mistral tests", arch: wormhole_b0, cmd: run_t3000_mistral_tests, timeout: 30, owner_id: U0896VBAKFC}, # Pratikkumar Prajapati
-          { name: "t3k qwen25_vl tests", arch: wormhole_b0, cmd: run_t3000_qwen25_vl_unit_tests, timeout: 40, owner_id: U07RY6B5FLJ}, #Gongyu Wang
+          { name: "t3k qwen25_vl tests", arch: wormhole_b0, cmd: run_t3000_qwen25_vl_unit_tests, timeout: 45, owner_id: U07RY6B5FLJ}, #Gongyu Wang
```
Comment (tt-rkim) [inline .github/workflows/t3000-unit-tests-impl.yaml]: @roseli-TT should look at this bump
We have 45 mins of unit tests for qwen 2.5? This one model?

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/sdpa_windowed.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+#include "ttnn/decorators.hpp"
+#include "ttnn/operations/transformer/sdpa_config.hpp"
+
+namespace ttnn {
+namespace operations::transformer {
+
+struct ExecuteWindowedScaledDotProductAttention {
+    static ttnn::Tensor invoke(
+        QueueId queue_id,
```
Comment (omilyutin-tt) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/sdpa_windowed.hpp:16]: I don't think this overload is needed

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_program_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+    log_debug(tt::LogOp, "q_chunk_size: {}", q_chunk_size);
+    log_debug(tt::LogOp, "k_chunk_size: {}", k_chunk_size);
+    log_debug(tt::LogOp, "q_num_chunks: {}", q_num_chunks);
+    log_debug(tt::LogOp, "k_num_chunks: {}", k_num_chunks);
+    log_debug(tt::LogOp, "NKH: {}", NKH);
+    log_debug(tt::LogOp, "cu_windows_seqlens.size(): {}", cu_window_seqlens_eles);
+
+    Program program = CreateProgram();
+
+    IDevice* device = input_tensor_q.device();
```
Comment (omilyutin-tt) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_program_factory.cpp]: Use `MeshDevice*` or `auto*`

**Review by TT-BrianLiu — COMMENTED** [review]
We already have some sdpa variants:
- ttnn/cpp/ttnn/operations/transformer/sdpa
   - Compute files: sdpa, joint_sdpa, ring_joint_sdpa
-  ttnn/cpp/ttnn/operations/transformer/sdpa_decode
   - Compute files: sdpa_flash_decode

These compute files I pointed out look largely the same to be (but probably have subtle important differences). Should we think about commonizing these files to avoid duplication? We can't fork this monster OP every time a new model comes out

File: .github/workflows/t3000-unit-tests-impl.yaml (hunk 1 above)
Comment (roseli-TT) [reply .github/workflows/t3000-unit-tests-impl.yaml]: https://github.com/tenstorrent/tt-metal/actions/runs/17041236450/job/48305988763
Passing run takes 23mins currently
@gwangTT you expect it to take up to 45mins?

**Comment by avoraTT** [comment]
> > Can this scheme be used in GPT OSS?
> 
> `windowed_SDPA` is made to work with image tokens and thus causality was unnecessary (thus not implemented). The code could be enhanced to support causality, which GPT OSS requires for causal sliding window attention.
> 
> @kpaigwar @avoraTT, what do you think?

Yes! GPT-OSS uses window attention in some of the layers, so it would be good to add support to the FlashAttention, as well as the FlashDecode, op. Currently, we have to generate a mask for every iteration.

File: .github/workflows/t3000-unit-tests-impl.yaml (hunk 1 above)
Comment (gwangTT) [reply .github/workflows/t3000-unit-tests-impl.yaml]: I made an effort to reduce the test time in a separate PR: https://github.com/tenstorrent/tt-metal/pull/27391. After that, I think we could be fine with the existing 40 minute of timeout value in face of the new unit test for windowed SDPA -- 7ed639ab896999e2d80f9330c2c5d4cef50ab8e7

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/sdpa_windowed.hpp (hunk 2 above)
Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/sdpa_windowed.hpp:16]: I tried removing this and the this code too: 
```c++
namespace transformer {

constexpr auto windowed_scaled_dot_product_attention = ttnn::register_operation<
    "ttnn::transformer::windowed_scaled_dot_product_attention",
    ttnn::operations::transformer::ExecuteWindowedScaledDotProductAttention>();

}  // namespace transformer
```

but got compilation error:
```
/localdev/gwang/tt-metal-2/ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/sdpa_windowed_pybind.cpp:49:55: error: no member named 'windowed_scaled_dot_product_attention' in namespace 'ttnn::transformer'; did you mean 'chunked_scaled_dot_product_attention'?
   49 |     using OperationType = decltype(ttnn::transformer::windowed_scaled_dot_product_attention);
      |                                    ~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/sdpa_windowed.hpp:16]: This does not seem like a overload to me. This is used to register the operation to TTNN as the quoted code above shows.

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_program_factory.cpp (hunk 3 above)
Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_program_factory.cpp]: 5f45693da6f3f4fca24147fcdb6a0a1221444c11

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/dataflow/reader_windowed.cpp [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+        }
+    };
+    auto get_window_indices = [&](uint32_t local_mask_idx) {
+        if (local_mask_idx < cu_window_seqlens_eles) {
+            auto low = get_cu_window_seqlens(local_mask_idx);
+            auto high = local_mask_idx == cu_window_seqlens_eles - 1 ? low : get_cu_window_seqlens(local_mask_idx + 1);
+            return std::make_pair(low, high);
+        }
+        auto low = get_cu_window_seqlens(cu_window_seqlens_eles - 1);
+        return std::make_pair(low, low);
```
[the remark this answers is not in the document]
Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/dataflow/reader_windowed.cpp:326]: At line 259, we are returning (low, high) tuple. Even though `high` could be set to `low` at line 258. The logic is still correct.

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/dataflow/reader_windowed.cpp [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+        std::min((local_q_start + q_chunks_per_core) * Sq_chunk_t, valid_Sqt) * tt::constants::TILE_HEIGHT;
+    uint32_t mask_windows_low_idx = 0;
+    bool found_mask_windows = false;
+    noc_async_read_barrier();  // Wait until reads are done
+    cb_push_back(cb_cu_window_seqlens_in, 1);
+    DPRINT_ARRAY_VIEW({
+        DPRINT << "cu_window_seqlens_eles: " << cu_window_seqlens_eles << ENDL();
+        DPRINT << "cu_window_seqlens: " << ENDL();
+        cb_cu_window_seqlens_ptr.print();
+    });
```
[the remark this answers is not in the document]
Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/dataflow/reader_windowed.cpp:341]: No need. `DPRINT_ARRAY_VIEW` has built-in switches on `WATCHER_OVERHEAD_OK`: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/array_view.hpp

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_program_factory.cpp [hunk 6]
```diff
@@ -0,0 +1,634 @@
...
+    log_debug(tt::LogOp, "log2_dht_granularity: {}", log2_dht_granularity);
+
+    // Reduce ops need to multiply by a scalar. We always want to multiply by 1.0f
+    class bfloat16 bfloat_identity_scalar(1.0f);
+    uint32_t packed_identity_scalar = pack_two_bfloat16_into_uint32({bfloat_identity_scalar, bfloat_identity_scalar});
+
+    union {
+        float f;
+        uint32_t u;
+    } scale_union;
```
[the remark this answers is not in the document]
Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_program_factory.cpp]: 2b2e225821c74e3d9982a2b1d234efc6d212079a

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/compute/sdpa_windowed.cpp [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+                for (uint32_t k_chunk = 0; (k_chunk * Sk_chunk_t) < q_high_idx; ++k_chunk) {
+                    const uint32_t k_low_idx = k_chunk * Sk_chunk_t;
+                    const uint32_t k_high_idx = k_low_idx + Sk_chunk_t;
+
+                    /* QK = Q_CHUNK @ K_CHUNK */
+                    pack_reconfig_data_format(cb_qk_im);
+                    // {
+                    //     const uint32_t& in0_cb = cb_q_in;
+                    //     const uint32_t& in1_cb = cb_k_in;
+                    //     const uint32_t& out_cb = cb_qk_im;
```
[the remark this answers is not in the document]
Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/compute/sdpa_windowed.cpp]: 1d7502012454c08a39653b083fa635de2fa87dae

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_op.cpp [hunk 8]
```diff
@@ -0,0 +1,231 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sdpa_windowed_op.hpp"
+#include <sys/types.h>
```
[the remark this answers is not in the document]
Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/sdpa_windowed_op.cpp]: 3cf04a9e3036a5bfdb9f43280a33b3846f2d0634

File: .github/workflows/t3000-unit-tests-impl.yaml (hunk 1 above)
Comment (gwangTT) [reply .github/workflows/t3000-unit-tests-impl.yaml]: I am running the unit tests pipeline again to make sure it is good: https://github.com/tenstorrent/tt-metal/actions/runs/17413968125 -- also added to the checklist in the description of this PR.

**Comment by gwangTT** [comment]
> We already have some sdpa variants:
> 
>     * ttnn/cpp/ttnn/operations/transformer/sdpa
>       
>       * Compute files: sdpa, joint_sdpa, ring_joint_sdpa
> 
>     * ttnn/cpp/ttnn/operations/transformer/sdpa_decode
>       
>       * Compute files: sdpa_flash_decode
> 
> 
> These compute files I pointed out look largely the same to be (but probably have subtle important differences). Should we think about commonizing these files to avoid duplication? We can't fork this monster OP every time a new model comes out

Agreed that there are common code among the sdpa variants. We should make an effort to refactor those common code into files that are shared among all the sdpa variants. I made [an issue](https://github.com/tenstorrent/tt-metal/issues/27797) to capture/track this work -- could be interesting to use AI to address that issue.

**Review by blozano-tt — APPROVED** [review]

File: .github/workflows/t3000-unit-tests-impl.yaml (hunk 1 above)
Comment (gwangTT) [reply .github/workflows/t3000-unit-tests-impl.yaml]: Done -- see the PR description's checklist.

File: ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/compute/compute_common.hpp [hunk 9]
```diff
@@ -0,0 +1,361 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
```
Comment (cglagovichTT) [inline ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/compute/compute_common.hpp]: I'd like to see compute_common.hpp and dataflow_common.hpp included from original SDPA since they are unchanged here.

Comment (gwangTT) [reply ttnn/cpp/ttnn/operations/transformer/sdpa_windowed/device/kernels/compute/compute_common.hpp]: 579f0ccd4c6d309a14f1da2c4998f5928482f390

**Review by cglagovichTT — APPROVED** [review]

**Review by sminakov-tt — APPROVED** [review]
